### DISCUSSION TITLE
add linear progress into ui.table's slot to reflect cell's value?

### DISCUSSION THREAD

**Discussion by retsyo** [zauberzeug/nicegui#2041] [Q&A] [answered]
### Question

In my simplified codes, the linear progress is used. The following is what I expected: 
- the `1st linear progress` has a postion at `10/(10+20)=66.67%`; while the 2nd one is `33.33%`
- whenever the `number` column changes, the two linear progress bars change. In my case, the total number(i.e. 30) is a constant which stands for total students in a class; so that if needed, we can pre-calculate the 2 percentages in advance in python. But it could be better if the table can calculate the total number and 2 percentages. 
- if possible, 2 numbers dispaly on `linear progress bar`, i.e. the percentage of taken, and the percentage of not-taken. But it is ok if only the percentage of taken is shown.

I don't know how to coin the code for add slot correctly. Can anyone help? Thanks
```python
from nicegui import ui

columns = [
    {'name': 'score', 'label': 'score', 'field': 'score'},
    {'name': 'number', 'label': 'number', 'field': 'number'},
    {'name': 'percent', 'label': 'percent', 'field': 'percent'},
]
rows = [
    {'score': '>=60', 'number': 10, 'percent':'_'*20},
    {'score': '<60', 'number': 20, 'percent':'_'*20},
]
tabRange = ui.table(columns=columns, rows=rows, row_key='name')
#~ tabRange.add_slot('body-cell-percent', r'''
#~ <q-tr :pros="pros">
    #~ <q-td key="sparkline" :pros="pros">
        #~ <q-linear-progress :value="{{ props.row.name }}" class="q-mt-md" />
    #~ </q-td>
#~ </q-tr>
#~ '''
#~ )

ui.run()
```

the image is a mockup one:
![mockup](https://github.com/zauberzeug/nicegui/assets/7960913/2c6a0b4f-5f45-4eaf-8730-ad800532efcc)

**Comment by retsyo** [comment]
no too much advance in my following code:
- I donno how to set value for individule `linear-progress`
- the `linear-progress` does not align with other cells in same row
- the `linear-progress` is trancated
- I donno how to put label on `linear-progress`
- I donno how to set color for `linear-progress`. But currently, blue for taken part, and gray for non-taken part, which is acceptable

```python
from nicegui import ui

columns = [
    {'name': 'score', 'label': 'score', 'field': 'score'},
    {'name': 'number', 'label': 'number', 'field': 'number'},
    {'name': 'percent', 'label': 'percent', 'field': 'percent'},
]
rows = [
    {'score': '>=60', 'number': 10, 'percent':'_'*20},
    {'score': '<60', 'number': 20, 'percent':'_'*20},
]
tabRange = ui.table(columns=columns, rows=rows, row_key='name')
with tabRange.add_slot('body-cell-percent'):
    ui.linear_progress().props('value=0.1').classes('w-full')

ui.run()
```

![tillnow](https://github.com/zauberzeug/nicegui/assets/7960913/19732d02-61b8-40de-b4ff-d7251520346d)

**Comment by falkoschindler** [comment, marked as answer]
Here is a starting point:
```py
columns = [
    {'name': 'score', 'label': 'score', 'field': 'score'},
    {'name': 'number', 'label': 'number', 'field': 'number'},
    {'name': 'percent', 'label': 'percent', 'field': 'percent'},
]
rows = [
    {'score': '>=60', 'number': 10, 'percent': 0.4},
    {'score': '<60', 'number': 20, 'percent': 0.6},
]
tabRange = ui.table(columns=columns, rows=rows, row_key='name')
tabRange.add_slot('body-cell-percent', r'''
    <q-td :props="props">
        <q-linear-progress size="1rem" :value="props.value" color="red" track-color="green">
            <div class="absolute-center text-black text-sm">
                {{ (props.value * 100).toFixed(0) }}%
            </div>
        </q-linear-progress>
    </q-td>
''')
```

- You had a typo "pros" instead of "props".
- I don't know how `'_'*20` was supposed to work.
- When defining a template that should generate cells based on the current cell value, you need to write a string instead of using `ui.linear_progress` directly.
